content/Phoenix for Rails developers - Bruno Vezoli.md

```diff
@@ -0,0 +1,347 @@
+
+### Helpers and Views
+
+Last but not least is how we present the data. You probably noticed in the file structure that there where two directories called `views/` and `templates/` on the Phoenix project structure, if you didn't go and take a closer view. You might be wondering what's the difference between them, well the templates are just that, text files with embedded code (just like erb we have eex) which you use to respond to requests. Views on the other hand resemble to helpers but with one noticeable difference: they are not global by default.
+
+Let's take a look at how they differ with an example; we would like to show the read time in our articles. Since we don't track this value on our DB and it's just for presentation you'd normally want to put this on a helper, like this:
+
+```ruby
+module ArticlesHelper
+  WORDS_PER_MINUTE = 200
+
+  def read_time(article)
+    words_count = article.text.split.size
+    read_time_in_minutes = words_count / WORDS_PER_MINUTE.to_f
+    “#{read_time_in_minutes.ceil} minutes"
+  end
+end
+```
+
+Easy peasy, now we just have to call this on our `show.html.erb` and we are done. Not so fast, the problem I see with this is that it gets included on every view. Sure for a small app like this and with a name like so specific there's no problem, but when it starts to grow and you have helpers with names like `format_title` and `format_title_for_user` you start having troubles. Phoenix's helpers (or views as they are called in the framework) are very similar:
+
+```elixir
+defmodule BlogWeb.ArticleView do
+  use BlogWeb, :view
+
+  alias Blog.Articles.Article
+
+  @words_per_minute 200
+
+  defp read_time(%Article{text: text}) do
+    words_count =
+      text
+      |> String.split(" ")
+      |> Enum.count()
+    read_time_in_minutes = Float.ceil(words_count / @words_per_minute)
+    "#{trunc(read_time_in_minutes)} minutes"
+  end
+end
+```
+
+But as I said before they are scoped to some presenter so this view, for example, has functions that can only be called on an article template. This allow you to use whatever name you want knowing you won't pollute the global scope. And what about if you want to use common functions across many templates? Simple, make another module with those functions and include it across the views where you need it.
+
+### Conclusion
+
+You might be thinking I threw a lot of dirt on Rails and the ways things are done, that I recommend not using Rails anymore and start using Phoenix. This is not one sof those "why X sucks and you should migrate all your code to Y", this is one of those "hey look at this cool new tool, you might find it useful/enjoyable". I still use Rails on a daily day basis and love the way it works, there are also many ways to mitigate some of the "bad" things about it, like presenters, services, form objects and many other stuff.
```

Comment:
`This is not one sof those "why X sucks and you should migrate all your code to Y"` sof?